streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -279,7 +282,7 @@ private long processNext(final List<ConsumerRecord<byte[], byte[]>> records,
             restorer.restoreBatchCompleted(offset + 1, records.size());
         }
 
-        return restoreConsumer.position(restorer.partition());
+        return consumerRecordIterator.hasNext() ? consumerRecordIterator.next().offset() : restoreConsumer.position(restorer.partition());
```

Comment:
Compare my comment for `StoreChangelogReaderTest` -- this seems to be not correct. I think we need to distinguish the case for which `offsetLimit == null` and `offsetLimit != null`

Reply:
I think this is still not correct, and as I synced with @mjsax and @bbejeck, currently we are returning the offset which is the skipped record's NEXT record's offset, instead of its own.

In fact @mjsax pointed out that the original code as in https://github.com/apache/kafka/pull/3948/files#diff-46ed6d177221c8778965ecb1b6657be3L263 is doing the right thing: if we break out of the current record batch, return the record's offset that is no smaller than the `end offset / offset limit` that triggers `hasCompleted`, otherwise return `consumer.position()` which returns the next available record's offset that already skip TC markers.

As of another potential issue I synced with @bbejeck , it turns out not an issue since when EoS is turned on, `consumer.endOffsets` already returns the offset of the next available record skipping the TC markers, i.e. it will be `LSO`. So we should be good by just reverting the changes in `processNext` and had the one-line fix on 242 above.